### PR TITLE
Add Python-interface for geometry constructor and geometry class

### PR DESCRIPTION
**Pull request by jorgensd**
Continuation of #3400.

Add `Geometry` Python class and appropriate constructor (`create_geometry`).
Expose some bindings and wrap return types (`Geometry` and `CoordinateElement`) when returned from Python objects.

### CODE REVIEW

File: python/dolfinx/fem/element.py [hunk 1]
```diff
@@ -46,6 +46,56 @@ def dtype(self) -> np.dtype:
...
+        The number of basis function is returned. E.g., for a linear
+        triangle cell the dimension will be 3.
+        """
+        return self._cpp_object.dim
+
+    def create_dof_layout(self) -> _cpp.fem.ElementDofLayout:
+        """Compute and return the dof layout"""
+        return self._cpp_object.create_dof_layout()
+
+    def push_forward(self, X: np.ndarray, cell_geometry: np.ndarray) -> np.ndarray:
```
Comment (jhale) [inline python/dolfinx/fem/element.py:62]: `npt` for typing?

File: python/dolfinx/fem/element.py [hunk 2]
```diff
@@ -46,6 +46,56 @@ def dtype(self) -> np.dtype:
...
+        """Compute the physical coordinates ``x`` of the reference coordinates ``X``
+        for the cell with coordinates ``cell_geometry``.
```
Comment (jhale) [inline python/dolfinx/fem/element.py:64]: As in `pull_back` I think the for... part is not needed.

File: python/dolfinx/fem/element.py [hunk 3]
```diff
@@ -46,6 +46,56 @@ def dtype(self) -> np.dtype:
...
+
+    @property
+    def variant(self) -> int:
+        """Return the Lagrange variant of the coordinate element.
+
+        Note:
+            Is returned as an integer and can be converted into a basix Lagrange variant using
+            ``basix.LagrangeVariant(value)``.
+        """
+        return self._cpp_object.variant
```
Comment (jhale) [inline python/dolfinx/fem/element.py:92]: I think it's OK to return the LagrangeVariant directly.

File: python/dolfinx/mesh.py [hunk 4]
```diff
@@ -64,13 +65,81 @@
...
+        """The ith element that describes the geometry map
+
+        Args:
+            i: The index of the element to return
+        """
+        return _CoordinateElement(self._cpp_object.cmaps(i))
+
+    @property
+    def dim(self):
+        """Euclidian dimension of the coordinate system"""
```
Comment (jhale) [inline python/dolfinx/mesh.py:107]: Dimension of the Euclidean coordinate system?

File: python/dolfinx/mesh.py [hunk 5]
```diff
@@ -819,3 +889,33 @@ def entities_to_geometry(
...
+    index_map: _IndexMap,
+    dofmap: npt.NDArray[np.int32],
+    element: _CoordinateElement,
+    x: np.ndarray,
+    input_global_indices: npt.NDArray[np.int64],
+) -> Geometry:
+    """Create a Geometry object
+
+    Args:
+        index_map: Index map associated with the geometry dofmap
```
Comment (jhale) [inline python/dolfinx/mesh.py:904]: full stop.

File: python/dolfinx/mesh.py [hunk 6]
```diff
@@ -819,3 +889,33 @@ def entities_to_geometry(
...
+        dofmap: The geometry (point) dofmap. For a cell, it gives the position
+            in the point array of each local geometry node
```
Comment (jhale) [inline python/dolfinx/mesh.py:906]: full stop.

Comment (jhale) [inline python/dolfinx/mesh.py:906]: position -> row
point array -> point coordinates ``x``

File: python/dolfinx/mesh.py [hunk 7]
```diff
@@ -819,3 +889,33 @@ def entities_to_geometry(
...
+        element: Element that describes the cell geometry map.
+        x: The point coordinates. The shape is ``(num_points, geometric_dimension)``
```
Comment (jhale) [inline python/dolfinx/mesh.py:908]: full stop

File: python/dolfinx/mesh.py [hunk 8]
```diff
@@ -64,13 +65,81 @@
...
+        """Euclidian dimension of the coordinate system"""
+        return self._cpp_object.dim
+
+    @property
+    def dofmap(self) -> npt.NDArray[np.int32]:
+        """Dofmap for the geometry, shape ``(num_cells, dofs_per_cell)``"""
+        return self._cpp_object.dofmap
+
+    def dofmaps(self, i: int) -> npt.NDArray[np.int32]:
+        """Return the ith dofmap
```
Comment (jhale) [inline python/dolfinx/mesh.py:116]: full stop - and throughout.

File: python/dolfinx/mesh.py [hunk 9]
```diff
@@ -64,13 +65,81 @@
...
+
+    def __init__(
+        self, geometry: typing.Union[_cpp.mesh.Geometry_float32, _cpp.mesh.Geometry_float64]
+    ):
+        """Initialize a geometry from a C++ geometry.
+
+        Args: The C++ geometry object
+
+        Note:
+            geometry objects should not usually be created using this
```
Comment (jhale) [inline python/dolfinx/mesh.py:86]: Caps Geometry.
Explain what user should do.

File: python/dolfinx/fem/element.py [hunk 10]
```diff
@@ -46,6 +46,56 @@ def dtype(self) -> np.dtype:
         """Scalar type for the coordinate element."""
         return np.dtype(self._cpp_object.dtype)
 
+    @property
+    def dim(self) -> int:
+        """The dimension of the coordinate element space.
+
+        The number of basis function is returned. E.g., for a linear
```
Comment (jhale) [inline python/dolfinx/fem/element.py:53]: E.g. -> e.g.

File: python/dolfinx/fem/element.py [hunk 11]
```diff
@@ -46,6 +46,56 @@ def dtype(self) -> np.dtype:
...
+
+    def push_forward(self, X: np.ndarray, cell_geometry: np.ndarray) -> np.ndarray:
+        """Compute the physical coordinates ``x`` of the reference coordinates ``X``
+        for the cell with coordinates ``cell_geometry``.
+
+        Args:
+            X: The physical coordinates of the reference points,
+                shape ``(num_points, topological_dimension)``.
+            cell_geometry: Physical coordinates describing the cell,
+                shape (:func:`dolfinx.fem.CoordinateElement.dim`, geometrical_dimension)``
```
Comment (jhale) [inline python/dolfinx/fem/element.py:70]: Why `:func:` ? Not done in line above.

File: python/dolfinx/fem/element.py [hunk 12]
```diff
@@ -46,6 +46,56 @@ def dtype(self) -> np.dtype:
...
+        """
+        return self._cpp_object.push_forward(X, cell_geometry)
+
+    def pull_back(self, x: np.ndarray, cell_geometry: np.ndarray) -> np.ndarray:
+        """Compute reference coordinates ``X`` for physical coordinates ``x``
+
+        Args:
+          x: Physical coordinates to pull back, shape ``(num_points, geometrical_dimension)``
+          cell_geometry: Physical coordinates describing the cell,
+              shape (:func:`dolfinx.fem.CoordinateElement.dim`, geometricaldimension)``
```
Comment (jhale) [inline python/dolfinx/fem/element.py:80]: `:func:` ?

**Review by jhale — CHANGES_REQUESTED** [review]

File: python/dolfinx/fem/element.py (hunk 10 above)
Comment (jorgensd) [inline python/dolfinx/fem/element.py:53]:
```suggestion
        The number of basis function is returned. e.g., for a linear
```

File: python/dolfinx/fem/element.py (hunk 1 above)
Comment (jorgensd) [reply python/dolfinx/fem/element.py:62]: Would be a bit nasty, as the typing would be `npt.NDArray[np.float32] | npt.NDArray[np.float64` as we can have either precision. What do you think?

File: python/dolfinx/fem/element.py (hunk 2 above)
Comment (jorgensd) [inline python/dolfinx/fem/element.py:64]:
```suggestion
        """Compute the physical coordinates ``x`` of the reference coordinates ``X``.
```

File: python/dolfinx/fem/element.py (hunk 12 above)
Comment (jorgensd) [inline python/dolfinx/fem/element.py:80]:
```suggestion
          cell_geometry: Physical coordinates describing the cell,
              shape ``(num_of_geometry_basis_functions, geometrical_dimension)``
```

File: python/dolfinx/fem/element.py (hunk 3 above)
Comment (jorgensd) [reply python/dolfinx/fem/element.py:92]: We do not return `basix` objects with type hints anywhere in DOLFINx, so I wasn't sure we would want to start here

File: python/dolfinx/fem/element.py [hunk 13]
```diff
@@ -46,6 +46,55 @@ def dtype(self) -> np.dtype:
...
+        return self._cpp_object.create_dof_layout()
+
+    def push_forward(self, X: np.ndarray, cell_geometry: np.ndarray) -> np.ndarray:
+        """Compute the physical coordinates ``x`` of the reference coordinates ``X``.
+
+        Args:
+            X: The physical coordinates of the reference points,
+                shape ``(num_points, topological_dimension)``.
+            cell_geometry: Physical coordinates describing the cell,
+                shape (:func:`dolfinx.fem.CoordinateElement.dim`, geometrical_dimension)``
```
Comment (jorgensd) [inline python/dolfinx/fem/element.py:69]:
```suggestion
                shape ``(num_geometry_basis_functions, geometrical_dimension)``
```

File: python/dolfinx/fem/element.py [hunk 14]
```diff
@@ -46,6 +46,55 @@ def dtype(self) -> np.dtype:
...
+        """Compute and return the dof layout"""
+        return self._cpp_object.create_dof_layout()
+
+    def push_forward(self, X: np.ndarray, cell_geometry: np.ndarray) -> np.ndarray:
+        """Compute the physical coordinates ``x`` of the reference coordinates ``X``.
+
+        Args:
+            X: The physical coordinates of the reference points,
+                shape ``(num_points, topological_dimension)``.
+            cell_geometry: Physical coordinates describing the cell,
```
Comment (garth-wells) [inline python/dolfinx/fem/element.py:68]: It would be helpful to describe `cell_geometry`. I'm not sure from the docstring what should be passed in.

**Review by garth-wells — COMMENTED** [review]
Are the pull-back/push-forward functions for single cells only? If yes, this will often be slow. Would be good to support  'vectorisation' over multiple cells.

File: python/dolfinx/fem/element.py (hunk 14 above)
Comment (jorgensd) [reply python/dolfinx/fem/element.py:68]: I am not sure what you would like to add here. 
It is the physical nodes/dofs of a given cell of the coordinate element.

File: python/dolfinx/mesh.py (hunk 9 above)
Comment (jorgensd) [inline python/dolfinx/mesh.py]:
```suggestion
            Geometry objects should not usually be created using this
```

File: python/dolfinx/mesh.py [hunk 15]
```diff
@@ -64,13 +65,82 @@
...
+
+    def __init__(
+        self, geometry: typing.Union[_cpp.mesh.Geometry_float32, _cpp.mesh.Geometry_float64]
+    ):
+        """Initialize a geometry from a C++ geometry.
+
+        Args: The C++ geometry object
+
+        Note:
+            geometry objects should usually be constructed with the :func:`create_geometry`
```
Comment (jorgensd) [inline python/dolfinx/mesh.py:86]:
```suggestion
            Geometry objects should usually be constructed with the :func:`create_geometry`
```

File: python/dolfinx/mesh.py [hunk 16]
```diff
@@ -819,3 +891,33 @@ def entities_to_geometry(
...
+
+
+def create_geometry(
+    index_map: _IndexMap,
+    dofmap: npt.NDArray[np.int32],
+    element: _CoordinateElement,
+    x: np.ndarray,
+    input_global_indices: npt.NDArray[np.int64],
+) -> Geometry:
+    """Create a Geometry object
```
Comment (jorgensd) [inline python/dolfinx/mesh.py:903]:
```suggestion
    """Create a Geometry object.
```

File: python/dolfinx/fem/element.py [hunk 17]
```diff
@@ -46,6 +46,55 @@ def dtype(self) -> np.dtype:
...
+
+        Note:
+            Is returned as an integer and can be converted into a basix Lagrange variant using
+            ``basix.LagrangeVariant(value)``.
+        """
+        return self._cpp_object.variant
+
+    @property
+    def degree(self) -> int:
+        """Return the degree of the coordinate element"""
```
Comment (jorgensd) [inline python/dolfinx/fem/element.py:95]:
```suggestion
        """Return the degree of the coordinate element."""
```

File: python/dolfinx/mesh.py [hunk 18]
```diff
@@ -64,13 +65,82 @@
...
+    """The geometry of a :class:`dolfinx.mesh.Mesh`"""
+
+    _cpp_object: typing.Union[_cpp.mesh.Geometry_float32, _cpp.mesh.Geometry_float64]
+
+    def __init__(
+        self, geometry: typing.Union[_cpp.mesh.Geometry_float32, _cpp.mesh.Geometry_float64]
+    ):
+        """Initialize a geometry from a C++ geometry.
+
+        Args: The C++ geometry object
```
Comment (jorgensd) [inline python/dolfinx/mesh.py:83]:
```suggestion
        Args: The C++ geometry object.
```

File: python/dolfinx/mesh.py [hunk 19]
```diff
@@ -64,13 +65,82 @@
...
+    @property
+    def cmap(self) -> _CoordinateElement:
+        """The element that describes the geometry map."""
+        return _CoordinateElement(self._cpp_object.cmap)
+
+    def cmaps(self, i: int) -> _CoordinateElement:
+        """The ith element that describes the geometry map.
+
+        Args:
+            i: The index of the element to return
```
Comment (jorgensd) [inline python/dolfinx/mesh.py:102]:
```suggestion
            i: The index of the element to return.
```

File: python/dolfinx/mesh.py [hunk 20]
```diff
@@ -64,13 +65,82 @@
...
+    @property
+    def dofmap(self) -> npt.NDArray[np.int32]:
+        """Dofmap for the geometry, shape ``(num_cells, dofs_per_cell)``."""
+        return self._cpp_object.dofmap
+
+    def dofmaps(self, i: int) -> npt.NDArray[np.int32]:
+        """Return the ith dofmap.
+
+        Args:
+            i: The index of the dofmap to return
```
Comment (jorgensd) [inline python/dolfinx/mesh.py:120]:
```suggestion
            i: The index of the dofmap to return.
```

**Comment by jorgensd** [comment]
> Are the pull-back/push-forward functions for single cells only? If yes, this will often be slow. Would be good to support 'vectorisation' over multiple cells.

This is it introducing new functionality and a non-backwards compatible API and is best addressed in a separate PR

File: python/dolfinx/fem/element.py (hunk 1 above)
Comment (jhale) [reply python/dolfinx/fem/element.py:62]: I think it's ok.

File: python/dolfinx/fem/element.py (hunk 3 above)
Comment (jhale) [reply python/dolfinx/fem/element.py:91]: Ok, good to be consistent.

**Review by jhale — APPROVED** [review]

**Comment by jorgensd** [comment]
@garth-wells 
I've made an issue describing potential actions that could be made in another PR, and some potential issues/interface changes one could make: https://github.com/FEniCS/dolfinx/issues/3408

File: cpp/dolfinx/mesh/Geometry.h [hunk 21]
```diff
@@ -113,7 +113,7 @@ class Geometry
   /// Move Assignment
   Geometry& operator=(Geometry&&) = default;
 
-  /// Return Euclidean dimension of coordinate system
+  /// Return Dimension of the Euclidean coordinate system
```
Comment (garth-wells) [inline cpp/dolfinx/mesh/Geometry.h:116]:
```suggestion
  /// Return dimension of the Euclidean coordinate system
```

File: python/dolfinx/fem/element.py [hunk 22]
```diff
@@ -46,6 +46,63 @@ def dtype(self) -> np.dtype:
         """Scalar type for the coordinate element."""
         return np.dtype(self._cpp_object.dtype)
 
+    @property
+    def dim(self) -> int:
+        """The dimension of the coordinate element space.
```
Comment (garth-wells) [inline python/dolfinx/fem/element.py:51]:
```suggestion
        """Dimension of the coordinate element space.
```

File: python/dolfinx/fem/element.py [hunk 23]
```diff
@@ -46,6 +46,63 @@ def dtype(self) -> np.dtype:
...
+
+        The number of basis function is returned. e.g., for a linear
```
Comment (garth-wells) [inline python/dolfinx/fem/element.py:53]:
```suggestion
        This is number of basis functions that span the coordinate space, e.g., for a linear
```

File: python/dolfinx/fem/element.py [hunk 24]
```diff
@@ -46,6 +46,63 @@ def dtype(self) -> np.dtype:
...
+    def create_dof_layout(self) -> _cpp.fem.ElementDofLayout:
+        """Compute and return the dof layout"""
+        return self._cpp_object.create_dof_layout()
+
+    def push_forward(
+        self,
+        X: typing.Union[npt.NDArray[np.float32], npt.NDArray[np.float64]],
+        cell_geometry: typing.Union[npt.NDArray[np.float32], npt.NDArray[np.float64]],
+    ) -> typing.Union[npt.NDArray[np.float32], npt.NDArray[np.float64]]:
+        """Compute the physical coordinates ``x`` of the reference coordinates ``X``.
```
Comment (garth-wells) [inline python/dolfinx/fem/element.py:67]:
```suggestion
        """Push points on the reference cell forward to the physical cell.
```

File: python/dolfinx/fem/element.py [hunk 25]
```diff
@@ -46,6 +46,63 @@ def dtype(self) -> np.dtype:
...
+
+        Args:
+            X: The physical coordinates of the reference points,
```
Comment (garth-wells) [inline python/dolfinx/fem/element.py:70]:
```suggestion
            X: Coordinates of points on the reference cell,
```

File: python/dolfinx/fem/element.py [hunk 26]
```diff
@@ -46,6 +46,63 @@ def dtype(self) -> np.dtype:
...
+                shape ``(num_points, topological_dimension)``.
```
Comment (garth-wells) [inline python/dolfinx/fem/element.py:71]:
```suggestion
                ``shape=(num_points, topological_dimension)``.
```

File: python/dolfinx/fem/element.py (hunk 14 above)
Comment (garth-wells) [reply python/dolfinx/fem/element.py:72]: Could could say something like 'geometry degree-of-freedom` and how the user would access the data in the right format.

File: python/dolfinx/fem/element.py [hunk 27]
```diff
@@ -46,6 +46,63 @@ def dtype(self) -> np.dtype:
...
+
+    def pull_back(
+        self,
+        x: typing.Union[npt.NDArray[np.float32], npt.NDArray[np.float64]],
+        cell_geometry: typing.Union[npt.NDArray[np.float32], npt.NDArray[np.float64]],
+    ) -> typing.Union[npt.NDArray[np.float32], npt.NDArray[np.float64]]:
+        """Compute reference coordinates ``X`` for physical coordinates ``x``.
+
+        Args:
+          x: Physical coordinates to pull back, shape ``(num_points, geometrical_dimension)``
```
Comment (garth-wells) [inline python/dolfinx/fem/element.py:85]:
```suggestion
          x: Physical coordinates to pull back to the reference cells, `` shape=(num_points, geometrical_dimension)``.
```

File: python/dolfinx/fem/element.py [hunk 28]
```diff
@@ -46,6 +46,63 @@ def dtype(self) -> np.dtype:
...
+          cell_geometry: Physical coordinates describing the cell,
```
Comment (garth-wells) [inline python/dolfinx/fem/element.py:86]: Explain how these can be accessed.

File: python/dolfinx/fem/element.py [hunk 29]
```diff
@@ -46,6 +46,63 @@ def dtype(self) -> np.dtype:
...
+              shape ``(num_of_geometry_basis_functions, geometrical_dimension)``
+        """
+        return self._cpp_object.pull_back(x, cell_geometry)
+
+    @property
+    def variant(self) -> int:
+        """Return the Lagrange variant of the coordinate element.
```
Comment (garth-wells) [inline python/dolfinx/fem/element.py:93]:
```suggestion
        """Lagrange variant of the coordinate element.
```

File: python/dolfinx/fem/element.py [hunk 30]
```diff
@@ -46,6 +46,63 @@ def dtype(self) -> np.dtype:
...
+
+        Note:
+            Is returned as an integer and can be converted into a basix Lagrange variant using
```
Comment (garth-wells) [inline python/dolfinx/fem/element.py:96]:
```suggestion
            The return type is an integer. A Basix enum can be crated using
```

File: python/dolfinx/fem/element.py [hunk 31]
```diff
@@ -46,6 +46,63 @@ def dtype(self) -> np.dtype:
...
+            ``basix.LagrangeVariant(value)``.
+        """
+        return self._cpp_object.variant
+
+    @property
+    def degree(self) -> int:
+        """Return the degree of the coordinate element."""
```
Comment (garth-wells) [inline python/dolfinx/fem/element.py:103]:
```suggestion
        """Polynomial degree of the coordinate element."""
```

File: python/dolfinx/mesh.py [hunk 32]
```diff
@@ -62,13 +63,82 @@
...
+    """The geometry of a :class:`dolfinx.mesh.Mesh`"""
+
+    _cpp_object: typing.Union[_cpp.mesh.Geometry_float32, _cpp.mesh.Geometry_float64]
+
+    def __init__(
+        self, geometry: typing.Union[_cpp.mesh.Geometry_float32, _cpp.mesh.Geometry_float64]
+    ):
+        """Initialize a geometry from a C++ geometry.
+
+        Args: The C++ geometry object.
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:81]: Use Google docstring style.

File: python/dolfinx/mesh.py [hunk 33]
```diff
@@ -62,13 +63,82 @@
...
+
+        Note:
+            Geometry objects should usually be constructed with the :func:`create_geometry`
+            and not this class initializer. This class is combined with different base classes
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:85]:
```suggestion
            and not using this class initializer. This class is combined with different base classes
```

File: python/dolfinx/mesh.py [hunk 34]
```diff
@@ -62,13 +63,82 @@
...
+            that depend on the scalar type used in the Geometry.
+        """
+
+        self._cpp_object = geometry
+
+    @property
+    def cmap(self) -> _CoordinateElement:
+        """The element that describes the geometry map."""
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:93]:
```suggestion
        """Element that describes the geometry map."""
```

File: python/dolfinx/mesh.py [hunk 35]
```diff
@@ -62,13 +63,82 @@
...
+        return _CoordinateElement(self._cpp_object.cmap)
+
+    def cmaps(self, i: int) -> _CoordinateElement:
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:96]: I don't think we need this yet. It is highly experimental. Best to remove for now.

File: python/dolfinx/mesh.py [hunk 36]
```diff
@@ -62,13 +63,82 @@
...
+    def dim(self):
+        """Dimension of the Euclidean coordinate system."""
+        return self._cpp_object.dim
+
+    @property
+    def dofmap(self) -> npt.NDArray[np.int32]:
+        """Dofmap for the geometry, shape ``(num_cells, dofs_per_cell)``."""
+        return self._cpp_object.dofmap
+
+    def dofmaps(self, i: int) -> npt.NDArray[np.int32]:
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:114]: Remove - experimental.

File: python/dolfinx/mesh.py [hunk 37]
```diff
@@ -62,13 +63,82 @@
...
+        """Return the ith dofmap.
+
+        Args:
+            i: The index of the dofmap to return.
+        """
+        return self._cpp_object.dofmaps(i)
+
+    def index_map(self) -> _IndexMap:
+        """The index map for the geometry."""
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:123]:
```suggestion
        """Index map describing the layout of the geometry points (nodes)."""
```

File: python/dolfinx/mesh.py [hunk 38]
```diff
@@ -62,13 +63,82 @@
...
+        return self._cpp_object.index_map()
+
+    @property
+    def input_global_indices(self) -> npt.NDArray[np.int64]:
+        """The global input indices of the geometry nodes."""
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:128]:
```suggestion
        """Global input indices of the geometry nodes."""
```

File: python/dolfinx/mesh.py [hunk 39]
```diff
@@ -62,13 +63,82 @@
...
+        return self._cpp_object.input_global_indices
+
+    @property
+    def x(self) -> typing.Union[npt.NDArray[np.float32], npt.NDArray[np.float64]]:
+        """The geometry data, with the shape ``(num_points, 3)``."""
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:133]:
```suggestion
        """Geometry coordinate points,  ``shape=(num_points, 3)``."""
```

File: python/dolfinx/mesh.py [hunk 40]
```diff
@@ -79,10 +149,12 @@ def __init__(self, mesh, domain: typing.Optional[ufl.Mesh]):
             domain: A UFL domain.
 
         Note:
-            Mesh objects should not usually be created using this
-            initializer directly.
+            Mesh objects should usually be constructed with the :func:`create_mesh`
+            and not this class initializer. This class is combined with different base classes
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:153]:
```suggestion
            and not using this class initializer. This class is combined with different base classes
```

File: python/dolfinx/mesh.py [hunk 41]
```diff
@@ -79,10 +149,12 @@ def __init__(self, mesh, domain: typing.Optional[ufl.Mesh]):
             domain: A UFL domain.
 
         Note:
-            Mesh objects should not usually be created using this
-            initializer directly.
+            Mesh objects should usually be constructed with the :func:`create_mesh`
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:152]:
```suggestion
            Mesh objects should usually be constructed using :func:`create_mesh`
```

File: python/dolfinx/mesh.py [hunk 42]
```diff
@@ -759,3 +831,33 @@ def entities_to_geometry(
...
+    index_map: _IndexMap,
+    dofmap: npt.NDArray[np.int32],
+    element: _CoordinateElement,
+    x: np.ndarray,
+    input_global_indices: npt.NDArray[np.int64],
+) -> Geometry:
+    """Create a Geometry object.
+
+    Args:
+        index_map: Index map associated with the geometry dofmap.
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:846]:
```suggestion
        index_map: Index map describing the layout of the geometry points (nodes).
```

File: python/dolfinx/mesh.py [hunk 43]
```diff
@@ -759,3 +831,33 @@ def entities_to_geometry(
...
+        dofmap: The geometry (point) dofmap. For a cell, it gives the row
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:847]: Add shape.

**Review by garth-wells — CHANGES_REQUESTED** [review]

**Review by garth-wells — APPROVED** [review]

**Pull request merged**